src/main/scala/ing/wbaa/druid/client/RequestFlowExtension.scala

```diff
@@ -0,0 +1,76 @@
+package ing.wbaa.druid.client
+
+import akka.http.scaladsl.model._
+import com.typesafe.config.{ Config, ConfigFactory }
+
+import scala.concurrent._
+
+/**
+  * Customization hook for altering the request flow in `DruidAdvancedHttpClient`. This is primarily
+  * intended to handle authentication requirements.
+  */
+trait RequestFlowExtension {
+
+  /**
+    * Modify the outgoing request before it is transmitted.
+    * @param request the outgoing request.
+    * @return the request that's to be sent to Druid.
+    */
+  def alterRequest(request: HttpRequest): HttpRequest
+
+  /**
+    * Modify the incoming response before it's decoded. This method can be implemented to intercept
+    * an incoming authentication challenge, reach out to an authentication system such as a token
+    * issuer, and then re-transmit the original request to Druid with appropriate authentication
+    * headers.
+    *
+    * Implementors should be aware that any request sent through the provided `druidEndpoint` will
+    * itself pass through the request flow extension. Be careful not to infinitely recurse.
+    *
+    * @param request the original, outgoing request.
+    * @param response the response that came back from the server.
+    * @param druidEndpoint invoke to transmit a request to Druid.
```

Comment:
I don't like `requestExecutor` because it's too generic. The provided function sends requests to (any of) the configured Druid broker(s). It cannot execute requests for anything else. If you want to send a request someplace else, you have to provide your own HTTP client to do the honours.

Reply:
I went with `sendToDruid`. This means implementors will end up writing the following:

```scala
sendToDruid(request.withHeaders(Authorization(...)))
```

I think that scans well enough. Certainly not worse than `requestExecutor`, while making explicit that the endpoint is Druid and not anything else.